### PR TITLE
Add GracefulShutdown function

### PR DESCRIPTION
**Pull request by justinpjose**
### What

**Describe what you have changed and why.**
Add Graceful Shutdown function in the main
Add GracefulShutdown Timeout in the config which also matches in the secrets of the API
Ensure the mongo connection is closed when closing the API

### How to review

**Describe the steps required to test the changes.**
Open a mongo connection
Run the API - "make debug" in the cmd
Close the API (Ctrl + C)
Check Mongo window - should inform the closing of the mongo connection

### Who can review

**Describe who worked on the changes, so that other people can review.**
Justin worked on these changes

### CODE REVIEW

File: cmd/dp-recipe-api/main.go [hunk 1]
```diff
@@ -36,87 +42,59 @@ func main() {
...
-	router.Methods("GET").Path("/recipes").HandlerFunc(recipeListHandler)
-	router.Methods("GET").Path("/recipes/{id}").HandlerFunc(recipeHandler)
-
-	log.Event(ctx, "starting http server", log.INFO, log.Data{"bind_addr": cfg.BindAddr})
-	srv := server.New(cfg.BindAddr, router)
-	if err := srv.ListenAndServe(); err != nil {
-		log.Event(ctx, "error starting http server for API", log.FATAL, log.Error(err))
+  
+	var err error
+	mongodb.Session, err = mongodb.Init()
```
Comment (CarlHembrough) [inline cmd/dp-recipe-api/main.go:53]: Should we be initialising Mongo if the feature flag is not enabled?

File: api/api.go [hunk 2]
```diff
@@ -0,0 +1,65 @@
+package api
```
Comment (CarlHembrough) [inline api/api.go:1]: There are lots of changes in this PR that are not related to the Shutdown function?

File: cmd/dp-recipe-api/main.go (hunk 1 above)
Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:53]: Initially that's what I did but then it became a problem when the gracefulShutdown was implemented. At first, the initialisation of `mongodb`was created within in the `if enableMongoData`. This means that `mongodb` would be only initialised when enabled.

But when implementing the gracefulShowdown, i cannot call the variable `mongodb` as i would get the error of mongodb being undefined in gracefulShowdown function. This is due to the fact that it is initialised in an if statement. 

That's why I coded in such a way that the initialised `mongodb`is only used when creating the RecipeAPI instance with Mongo in datastore when enabled

Is there another way in which I can achieve it?

File: api/api.go (hunk 2 above)
Comment (justinpjose) [reply api/api.go:1]: They are layered branches in which it contains all the commits from the previous PRs as well.

File: cmd/dp-recipe-api/main.go (hunk 1 above)
Comment (CarlHembrough) [reply cmd/dp-recipe-api/main.go:53]: It should be fine to move the initialisation into the `if enableMongoData`block. If the reference to `mongodb` still lives outside of the if statement, and you also have the if statement in the graceful shutdown (which you already do). Give it a try and let me know if you have any issues 👍

File: api/api.go (hunk 2 above)
Comment (nshumoogum) [reply api/api.go:1]: Thats why as a reviewer I dont think we should be looking at these pull requests until the previous prs are reviewed, updated, approved and merged

Comment (nshumoogum) [reply api/api.go:1]: Or change the base branch to the branch that this branch was created off - that way the diff will be between the two branches which actually care about at this point in the git history

Comment (justinpjose) [reply api/api.go]: Changing the base branch did the trick. Thank you

File: cmd/dp-recipe-api/main.go (hunk 1 above)
Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:53]: Done in commit "Initialising mongo session within feature flag"

File: cmd/dp-recipe-api/main.go [hunk 3]
```diff
@@ -41,18 +43,19 @@ func main() {
...
+	mongodb := &mongo.Mongo{
+		Collection: cfg.MongoConfig.Collection,
+		Database:   cfg.MongoConfig.Database,
+		URI:        cfg.MongoConfig.BindAddr,
+	}
 
-		var err error
-		mongodb.Session, err = mongodb.Init()
+	if enableMongoData {
+		session, err := mongodb.Init()
```
Comment (CarlHembrough) [inline cmd/dp-recipe-api/main.go:53]: I think you can still assign directly to `mongodb.Session` here:
```
mongodb.Session, err = mongodb.Init()
``` 
this will remove the need for the else statement?

File: cmd/dp-recipe-api/main.go [hunk 4]
```diff
@@ -64,4 +67,36 @@ func main() {
 		api.CreateAndInitialiseRecipeAPI(ctx, *cfg, store.DataStore{Backend: nil})
 	}
 
+	apiErrors := make(chan error, 1)
+
+	// Gracefully shutdown the application closing any open resources.
+	gracefulShutdown := func() {
+		log.Event(ctx, (fmt.Sprintf("shutdown with timeout: %s", cfg.GracefulShutdownTimeout)), log.INFO)
```
Comment (CarlHembrough) [inline cmd/dp-recipe-api/main.go:74]: no need for the braces around `fmt.Sprintf`

File: cmd/dp-recipe-api/main.go [hunk 5]
```diff
@@ -64,4 +67,36 @@ func main() {
...
+		ctx, cancel := context.WithTimeout(context.Background(), cfg.GracefulShutdownTimeout)
```
Comment (CarlHembrough) [inline cmd/dp-recipe-api/main.go:75]: The usage for the shutdown context timeout is not quite right. I would expect the context to be declared outside of the go routine, and then a wait on the done channel `<-ctx.Done()` 

See here for an example: https://github.com/ONSdigital/dp-dataset-api/blob/develop/main.go#L167

File: cmd/dp-recipe-api/main.go [hunk 6]
```diff
@@ -64,4 +67,36 @@ func main() {
...
+		log.Event(ctx, "shutdown complete", log.INFO)
+
+		cancel()
+		os.Exit(1)
+	}
+
+	for {
+		select {
+		case err := <-apiErrors:
+			log.Event(ctx, "api error received", log.ERROR, log.Error(err))
```
Comment (CarlHembrough) [inline cmd/dp-recipe-api/main.go:96]: should we also do a graceful shutdown if an API error occurs?

File: cmd/dp-recipe-api/main.go (hunk 3 above)
Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:53]: If I initialise `err` before that, then I can do the above. Therefore, I have done that and committed in "Simplifying code and improving gracefulShutdown."

File: cmd/dp-recipe-api/main.go (hunk 4 above)
Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:74]: Done in commit "Simplifying code and improving gracefulShutdown."

Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:74]: Done in PR https://github.com/ONSdigital/dp-recipe-api/pull/127

File: cmd/dp-recipe-api/main.go (hunk 6 above)
Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:96]: When I was doing the graceful Shutdown with Nathan, we also look for apiErrors and then ensure that all services are closed gracefully even if the apiError occurred for a specific reason. This is the same in the DatasetAPI as well.

File: cmd/dp-recipe-api/main.go (hunk 5 above)
Comment (justinpjose) [reply cmd/dp-recipe-api/main.go:75]: Shutdown Context Timeout successfully Done in PR https://github.com/ONSdigital/dp-recipe-api/pull/127
